### DISCUSSION TITLE
Como bugar o GitHub - Introdução ao Git Internals

### DISCUSSION THREAD

**Discussion by felquis** [frontendbr/forum#305] [Dicas]
Hoje aprendi uma coisa, que não tem muita utilizada, mas é um conhecimento interessante.

No inicio o GIT era um toolkit para criar Version Control Systems (VCS), este toolkit contem vários comandos de baixo niveis que são categorizados como Plumbing Commands, mais tarde foi criado comandos mais user-friendly que foram categorizado como Porcelain Commands que são os que usamos no dia-a-dia como: init, add, commit, push, pull, branch, tag, merge, etc... 

> Chapter [Git Internals Plumbing and Porcelain](https://git-scm.com/book/tr/v2/Git-Internals-Plumbing-and-Porcelain)

A litghning talk em que aprendi isso foi a ["Talk #2 - Digging deep into Git"](https://www.youtube.com/watch?v=H2j7e81J798) o @greenboxal fala sobre bugar o GitHub em [19m43s](https://youtu.be/H2j7e81J798?t=19m43s)
## Como bugar o GitHub?

**Passo 1** criar a hash de um conteúdo 

```
echo "Secret Of Life" | git hash-object -w --stdin
```

output: 68f72a82f53ad0f0854dc9259521a2a9a8222a20

**Passo 2** criar uma tag que linca para a hash acima

```
git update-ref refs/tags/tag_name 68f72a82f53ad0f0854dc9259521a2a9a8222a20
```

**Passo 3** Criar um README.md ou .gitignore e fazer push para o GitHub

**Passo 4** Olhar o resultado bizarro no GitHub

Primeiro, que o Secret Of Life não está em nenhum local visivel no GitHub

Ele diz que tenho 1 release 
![image](https://cloud.githubusercontent.com/assets/736728/18634819/040fb9be-7e59-11e6-8ba0-1302d7fe8ab7.png)
Mas se entro na página de releases diz que não tenho nenhum
![image](https://cloud.githubusercontent.com/assets/736728/18634833/1301040a-7e59-11e6-9bd3-b1f29fb3993a.png)

Diz que tenho uma branch secret
![image](https://cloud.githubusercontent.com/assets/736728/18634838/1c21378a-7e59-11e6-87d8-00562815bcac.png)
Mas se clico na branch vou para uma página 404 https://github.com/felquis/secret/tree/secret
## Conclusão

Git Internals é divertido porque você pode confundir o GitHub, claro que não tem utilidade nenhuma confundir o GitHub, mas é interessante para instigar ir a fundo porque quanto mais a gente sabe da ferramenta que usamos, mais conscientes somos do que estamos de fato fazendo :)
## El Gran Final

Coloquei o que eu acho que é o segredo da vida no repositório https://github.com/felquis/secret claro que no GitHub vc não vai conseguir ler este segredo, para ler use.

```
git clone git@github.com:felquis/secret.git
cd secret
git cat-file -p secret
```

Espero que você tenha se divertido! E compartilhe mais coisas loucas sobre Git Internals com a gente :D :frog::ok_hand:

**Comment by fdaciuk** [comment]
Bem legal @felquis haha :+1:

**Comment by ghost** [comment]
Baixei só pra descobrir qual o segredo da vida pra você. hahaha

**Comment by felipefialho** [comment]
@felquis Concordo com sua definição de segredo da vida 🌗

**Comment by fredericksilva** [comment]
Incrível 👍

**Comment by fabiomirandaa** [comment]
Se fosse um 42 ia ser até mais engraçado.

**Comment by nicklima** [comment]
hahaha. Muito bom!

**Comment by DavidCardoso** [comment]
Todo mundo deveria conhecer e usar esse segredo! ;)
